snuba/datasets/transactions_processor.py

```diff
@@ -44,7 +44,7 @@ class TransactionsMessageProcessor(MessageProcessor):
     }
 
     def __extract_timestamp(self, field):
-        timestamp = _ensure_valid_date(datetime.fromtimestamp(field))
+        timestamp = _ensure_valid_date(datetime.utcfromtimestamp(field))
```

Comment:
Any reason for not using `fromtimestamp` providing the appropriate timezone as recommended on the docs instead of utcfromtimestamp ?
https://docs.python.org/3/library/datetime.html#datetime.datetime.utcfromtimestamp

Reply:
Ah no I didn't realize there was a recommended way. I'll change it.

Reply:
Be careful with this: [`JSONRowEncoder` doesn't include the timestamp either when writing records](https://github.com/getsentry/snuba/blob/f6660b21ba74ed43b8a608ebf772f8e5d4c6aae6/snuba/clickhouse/http.py#L30) per `DATETIME_FORMAT`: https://github.com/getsentry/snuba/blob/7bfa538516882dc58b2bed2f3e93e87360a6719c/snuba/clickhouse/__init__.py#L1 (Not saying it should be this way, or that it can't be changed.)

In this case both `fromtimestamp` with a timezone and `utcfromtimestamp` would be treated equivalently with respect to what is sent to ClickHouse, both of which will result in incorrect results if your local timestamp is inconsistent with the ClickHouse server timestamp:

```
>>> ts = datetime.now().timestamp()
>>> datetime.fromtimestamp(ts).strftime(DATETIME_FORMAT)
'2020-09-17 21:08:01'
>>> datetime.fromtimestamp(ts, pytz.utc).strftime(DATETIME_FORMAT)
'2020-09-18 04:08:01'
>>> datetime.utcfromtimestamp(ts).strftime(DATETIME_FORMAT)
'2020-09-18 04:08:01'
```

Reply:
I'm going to leave this as is:
1) You can't subtract offset-aware and naive datetimes, which means that `_ensure_valid_date` throws an exception if I use `fromtimestamp(ts, pytz.utc)`. I can change `_ensure_valid_date` to use an offset-aware datetime, but then that breaks for all other uses of that helper function.
2) This isn't an issue in all the other processors, which use the `event["datetime"]` field to create a naive UTC datetime. This issue crops up in transactions only because `datetime.fromtimestamp` assumes the timestamp is in the local tz and transforms the datetime accordingly. This change puts the transaction processor in the same place as the other processors.
3) We can afford to use naive datetimes in snuba because all the ingestion pipelines sanitize data to be UTC before they get to snuba.

Reply:
It seems to me that unless both clickhouse and Snuba are both consistently on UTC time zone both the current situation and the fix would be broken in different ways.

This is the scenario both with or without this fix.
1) snuba processor receives a timestamp that is sanitized to UTC (let's hope relay is running on utc as well)
2) snuba stores a naive timestamp in clickhouse 
3) clickhosue interprets it to whatever its time zone is
so far everything may be ok
4) when we query, depending on the code path, we inconsistently compare dates either with `toDateTime('XXXX')` or `toDateTime('XXXXX', 'Universal')`. Depending whether the datetime was properly recognized as a datetime by the parser. This is the same problem regarding date time literals we discussed before.
5) Now clickhouse will compare the naive datetime it ingested with the expression in the query and one of the following is possible (assuming sentry provides datetimes  in queries that are supposed to be sanitized to UTC).
5a) clickhouse is on UTC, `toDateTime('XXX')` in the query -> All good. naive compared with naive on the same time zone
5b) clickhouse is on UTC, `toDateTime('XXX', 'Universal')` -> All good. naive (on a UTC system) compared with UTC.
5c) clickhouse is not on UTC, `toDateTime('XXX')` in the query -> All good by chance. naive compared with naive on a system that is not on UTC
5d) clickhouse is not on UTC, `toDateTime('XXX', 'Universal')` -> Broken. naive data stored (interpreted according to clickhouse time zone) compared with something clickhouse translates into utc.

It seems this system is fundamentally broken no matter on this PR. 
Am I getting this right, or am I missing something ?

Reply:
If this makes sense, we got reports of issues when Clickhouse was not running on UTC from self-hosted users.

Reply:
I think the purpose of this PR is getting lost. This is not meant to solve all datetime/timezone issues in Snuba right now. This is meant to solve the timezone inconsistency between transactions and other events. 

The issue right now is: Snuba is running in local time. A transaction event is sent to it. The timestamp for the transaction data will be different from that of the event data because Snuba is parsing the transaction timestamps as local time instead of UTC time. Now the transaction and event data are out of sync.

Once this change goes through, we can solve the naive/aware issue in Snuba if we want. Or we can add a line to our documentation that says "run snuba/clickhouse in UTC to avoid timezone errors". Either way those solutions are out of scope for this work.

Reply:
ok, makes sense. 
Mind adding a task on the backlog to revisit this. I think this should still be considered an open problem.

Before I accept, my original question for @BYK: this can change the current behavior for those self hosted users who are not running snuba on UTC. Should we do something about that? Like forcing snuba to run in UTC from the next version? Provide some docs about this behavior? Allow users to switch from the old to the new ?
Knowing that if they are running snuba on a different time zone than UTC their errors and transactions are already behaving inconsistently.

Reply:
I couldn't find definitive documentation anywhere but I strongly believe all Docker images run on UTC by default so this should be fine for self-hosted. We may want to add a note to the README for those rare people that try to run Snuba differently.

Reply:
I added a line to the README and a task to the backlog.